# BitsxlaMarato 2023: Hacking the secrets of the seminal microbiome

## Overview
This repository contains the codebase and documentation for our award-winning project developed during the [BitsxlaMarat√≥ 2023 Hackathon](https://www.fib.upc.edu/ca/la-marato) at the Faculty of Informatics of Barcelona (FIB), Technical University of Catalonia (UPC). The goal of our project is to investigate potential associations between bacterial profiles, derived from clinical and seminal microbiota data of infertile patients and controls, and male fertility. We have employed advanced bioinformatics tools, statistical analysis, and machine learning algorithms for this purpose.

## Hackathon Victory
üèÜ **We Won the Hackathon!** Our project was awarded first place at the [BitsxlaMarat√≥ 2023 Hackathon](https://www.fib.upc.edu/ca/la-marato) organized by the Faculty of Informatics of Barcelona (FIB), [Hackers@UPC](https://hackersatupc.org/) and [Barcelona Supercomputing Center - Centro Nacional de Supercomputaci√≥n (BSC](https://www.bsc.es/). We are excited and grateful for this recognition, highlighting our team's dedication and innovative approach in investigating potential associations between bacterial profiles and male fertility.

To delve into our project's journey, check out our [Devpost submission](<link-to-your-devpost-submission>).

## Project Structure
data: Contains the datasets used for analysis.
notebooks: Jupyter notebooks for data analysis, PCA, clustering, and modeling.
src: Python source code for the backend of the web application.
web: Frontend code for the web application.
docs: Documentation files, including this README.


    ‚îú‚îÄ‚îÄ LICENSE
    ‚îú‚îÄ‚îÄ README.md          <- The top-level README for developers using this project.
    ‚îú‚îÄ‚îÄ requirements.txt         
    ‚îú‚îÄ‚îÄ data
    ‚îÇ¬†¬† ‚îú‚îÄ‚îÄ raw            <- The original, immutable data dump.
    ‚îÇ   ‚îÇ   ‚îú‚îÄ‚îÄ read_data.ipynb <- Script to read input xlsx files and generate a csv for each page.
    ‚îÇ   ‚îÇ¬†¬† ‚îú‚îÄ‚îÄ illumina <- Data obtained by Illumina sequencing.
    ‚îÇ   ‚îÇ   ‚îÇ   ‚îú‚îÄ‚îÄ Family-level microbiota.csv
    ‚îÇ   ‚îÇ   ‚îÇ   ‚îú‚îÄ‚îÄ Genus-level microbiota.csv
    ‚îÇ   ‚îÇ   ‚îÇ   ‚îú‚îÄ‚îÄ Pylum-level microbiota.csv
    ‚îÇ   ‚îÇ   ‚îÇ   ‚îú‚îÄ‚îÄ Sample info + Sperm quality.csv
    ‚îÇ   ‚îÇ   ‚îÇ   ‚îú‚îÄ‚îÄ README.txt
    ‚îÇ   ‚îÇ   ‚îÇ   ‚îî‚îÄ‚îÄ Metadata_and_relative_abundance_of_seminal_microbiota_from_idiopathic_infertile_patients_and_donors.xlsx
    ‚îÇ   ‚îÇ   ‚îú‚îÄ‚îÄ nanopore <- Data obtained by Nanopore sequencing.
    ‚îÇ   ‚îÇ   ‚îÇ   ‚îú‚îÄ‚îÄ MANIFEST.TXT
    ‚îÇ   ‚îÇ   ‚îÇ   ‚îú‚îÄ‚îÄ README.txt
    ‚îÇ   ‚îÇ   ‚îÇ   ‚îî‚îÄ‚îÄ Relative abundance of seminal microbiota from idiopathic infertile patients and donors using MinION sequencing.xlsx
    ‚îÇ¬†¬† ‚îÇ   ‚îî‚îÄ‚îÄ NP_metabolica <- Metabolite-Microbe database.
    ‚îÇ   ‚îÇ       ‚îú‚îÄ‚îÄ NPAtlas.csv
    ‚îÇ   ‚îÇ       ‚îî‚îÄ‚îÄ NPAtlas.xlsx
    ‚îÇ   ‚îÇ
    ‚îÇ   ‚îî‚îÄ‚îÄ processed  
    ‚îÇ       ‚îú‚îÄ‚îÄ family_clusters_fertility.csv
    ‚îÇ       ‚îú‚îÄ‚îÄ genus_clusters_fertility.csv
    ‚îÇ       ‚îî‚îÄ‚îÄ pylum_clusters_fertility.csv
    ‚îÇ  
    ‚îú‚îÄ‚îÄ src
    ‚îÇ¬†¬† ‚îú‚îÄ‚îÄ features
    ‚îÇ   ‚îÇ   ‚îî‚îÄ‚îÄ read_data.ipynb
    ‚îÇ   ‚îÇ
    ‚îÇ   ‚îî‚îÄ‚îÄ models 
    ‚îÇ       ‚îú‚îÄ‚îÄ KNN 
    ‚îÇ       ‚îÇ   ‚îú‚îÄ‚îÄ family_knn
    ‚îÇ       ‚îÇ   ‚îú‚îÄ‚îÄ genus_knn
    ‚îÇ       ‚îÇ   ‚îî‚îÄ‚îÄ pylum_knn
    ‚îÇ       ‚îú‚îÄ‚îÄ MLP 
    ‚îÇ       ‚îÇ   ‚îú‚îÄ‚îÄ family_mlp
    ‚îÇ       ‚îÇ   ‚îú‚îÄ‚îÄ genus_mlp
    ‚îÇ       ‚îÇ   ‚îî‚îÄ‚îÄ pylum_mlp
    ‚îÇ       ‚îî‚îÄ‚îÄ RandomForest 
    ‚îÇ           ‚îú‚îÄ‚îÄ family_random_forest
    ‚îÇ           ‚îú‚îÄ‚îÄ genus_random_forest
    ‚îÇ           ‚îî‚îÄ‚îÄ pylum_random_forest
    ‚îÇ    
    ‚îî‚îÄ‚îÄ analysis
     ¬†¬† ‚îú‚îÄ‚îÄ KNN
        ‚îÇ   ‚îú‚îÄ‚îÄ family_clusters_knn.ipynb
        ‚îÇ   ‚îú‚îÄ‚îÄ genus_clusters_knn.ipynb
        ‚îÇ   ‚îî‚îÄ‚îÄ pylum_clusters_knn.ipynb
     ¬†¬† ‚îú‚îÄ‚îÄ MLP
        ‚îÇ   ‚îú‚îÄ‚îÄ family_mlp_model.ipynb
        ‚îÇ   ‚îú‚îÄ‚îÄ genus_mlp_model.ipynb
        ‚îÇ   ‚îî‚îÄ‚îÄ pylum_mlp_model.ipynb
     ¬†¬† ‚îú‚îÄ‚îÄ RandomForest
        ‚îÇ   ‚îú‚îÄ‚îÄ family_randomforest_model.ipynb
        ‚îÇ   ‚îú‚îÄ‚îÄ genus_randomforest_model.ipynb
        ‚îÇ   ‚îî‚îÄ‚îÄ pylum_randomforest_model.ipynb
        ‚îú‚îÄ‚îÄ analysis_clinical_info.ipynb
        ‚îî‚îÄ‚îÄ check_fertility_model_pylum.ipynb
  
## Web Application
We've crafted a user-friendly web application for doctors to input samples and receive predictions on cluster membership and fertility status. The app also features visualizations highlighting new samples within overall cluster distributions. Currently in the final development stage, the web page is accessible in the corresponding repository branch because it is not fully functional yet. Your feedback is welcomed as we add the finishing touches before the official launch. Stay tuned for an enhanced user experience in microbial sample analysis and fertility assessment.

## Contributors
The people involved in this project are [Arlet Corominas](https://github.com/arletcoro), [Alexandra Gonz√°lez](https://github.com/alexandraglz) and [Hakima Marouan](https://github.com/haakima).

## Acknowledgments
We would like to express our gratitude to [La Marat√≥ de TV3](https://www.ccma.cat/tv3/marato/2023/310/) and the [Faculty of Informatics at UPC](https://www.fib.upc.edu) for organizing this hackathon and providing us with the opportunity to work on this meaningful project.

## License
This project is licensed under the [License Name] - see the LICENSE.md file for details.

## Contact
For questions or inquiries, please contact:

- [Arlet Corominas](mailto:arlet.corominas@estudiantat.upc.edu)
- [Alexandra Gonz√°lez](mailto:alexandra.gonzalez.alvarez@estudiantat.upc.edu)
- [Hakima Marouan](mailto:hakima.marouan@estudiantat.upc.edu)
